ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/mainnet/ProtocolSpecBuilder.java

```diff
@@ -74,7 +76,8 @@
     return this;
   }
 
-  public ProtocolSpecBuilder<T> evmBuilder(final Function<GasCalculator, EVM> evmBuilder) {
+  public ProtocolSpecBuilder<T> evmBuilder(
+      final BiFunction<GasCalculator, MetricsSystem, EVM> evmBuilder) {
```

Comment:
This is why I think we need a DI system.  The Function/BiFunction/TriFuction/... becomes a Supplier<> and the context is provided by the DI system.

Reply:
The point of this ugliness is that you can take a `ProtocolSpecBuilder` configured for one milestone and change some of the inputs to adapt it to either a different milestone or for something like IBFT or Clique.  So Frontier defines a evmBuilder which is unchanged in homestead, but homestead provides a different GasCalculator input. I imagine you can make a DI system achieve this but I'd expect it to be a fairly advanced use case that winds up being complex either in how it's written or how deep an understanding of the DI library you need.

It is another one of those things where at some point we need to just try it and see how it actually comes out though.